Assets/Mirage/ObjectPolling/ObjectPoolManager.cs

```diff
@@ -0,0 +1,204 @@
+
+        /// <summary>
+        ///     Spawn handler for all network spawning of things.
+        /// </summary>
+        /// <param name="msg">The message incoming from spawn handler.</param>
+        /// <returns>Return back the network identity of the object needing soawning.</returns>
+        internal NetworkIdentity SpawnObject(SpawnMessage msg)
+        {
+            NetworkIdentity spawnedObject = NetworkSpawnPool(msg.position, Quaternion.identity, msg.assetId, 1);
+
+            // Active the pooled object.
+            spawnedObject.gameObject.SetActive(true);
+
+            // return the game object back to mirage to have it finish spawning.
+            return spawnedObject;
+        }
+
+        /// <summary>
+        ///     Spawn handler for all network unspawning of things.
+        /// </summary>
+        /// <param name="spawned">What network identity we are going to unspawn and return back to network pool.</param>
+        internal void UnSpawnObject(NetworkIdentity spawned)
+        {
+            ObjectPooling<NetworkIdentity> p = default;
+
+            foreach (object pool in _recycledObjects.Values)
+            {
+                if (!(pool is ObjectPooling<NetworkIdentity> obj) || !obj.PooledObjectIDs.Contains(spawned.gameObject.GetInstanceID())) continue;
+
+                p = obj;
+
+                break;
+            }
+
+            p.Despawn(spawned);
+        }
+
+        /// <summary>
+        ///     Unspawn of local gameobjects
+        /// </summary>
+        /// <param name="spawned">What object we are going to unspawn and return back to local pool.</param>
+        private void UnSpawnObject<T>(T spawned) where T : Object
+        {
+            ObjectPooling<T> p = default;
+
+            foreach (object pool in _recycledObjects.Values)
+            {
+                if (!(pool is ObjectPooling<T> obj) || !obj.PooledObjectIDs.Contains(spawned.GetInstanceID())) continue;
+
+                p = obj;
+
+                break;
+            }
+
+            p.Despawn(spawned);
+        }
+
+        #endregion
+
+        #region Network Spawning
+
+        /// <summary>
+        ///     Spawn an object using our pooled objects for the specific object
+        ///     or create a new pool if no pool has been created yet.
+        /// </summary>
+        /// <param name="position">The position we want to spawn the object at.</param>
+        /// <param name="rotation"></param>
+        /// <param name="AssetId">The asset id of the object from <see cref="NetworkIdentity" /> to use to spawn object of.</param>
+        /// <param name="quantity">How many pooled objects we want to create of this object. If we don't have a pool started.</param>
+        /// <returns></returns>
+        internal NetworkIdentity NetworkSpawnPool(Vector3 position, Quaternion rotation, Guid AssetId, int quantity = 3)
+        {
+            int prefabId = _objectsAssetIds[AssetId].gameObject.GetInstanceID();
```

Comment:
why do we need to use GetInstanceID here?

Reply:
Looks like this is using a key (AssetId) to find another key (prefabId)? can we just use `AssetId` instead?

Reply:
No because it needs to look up the prefab that has been used for instantiate... Each prefab that gets spawned for pooling is different id we cannot use that.